Possibility to define multiple inline-secrets for mutation

kschu91 opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes #1001 
| License         | Apache 2.0


### What's in this PR?
https://github.com/banzaicloud/bank-vaults/issues/1001#issuecomment-707641069


### Why?
https://github.com/banzaicloud/bank-vaults/issues/1001#issuecomment-707641069

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)

### To Do
<!-- (Please remove this section if you don't need it.) -->
- [x] Please review implementation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/banzaicloud/bank-vaults?pullRequest=1143) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/banzaicloud/bank-vaults?pullRequest=1143) before we can accept your contribution.<br/><hr/>**Niklas Faßbender** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/banzaicloud/bank-vaults?pullRequest=1143) it.</sub>

bonifaido commented:
Hi @kschu91, regarding the failed build here: https://github.com/banzaicloud/bank-vaults/pull/1143/checks?check_run_id=1252775994

Could you please add the required ignore line to the license check config? https://github.com/banzaicloud/bank-vaults/blob/master/.licensei.toml#L27

bonifaido commented:
Don't worry about my previous comment, just added it in https://github.com/banzaicloud/bank-vaults/pull/1144/commits/e2f8a3566803dda479b6ced57c1d218d89457938

Please rebase on master once this gets merged (a few minutes from now on).

bonifaido commented on cmd/vault-secrets-webhook/common.go:
```diff
@@ -42,3 +42,7 @@ func getDataFromVault(data map[string]string, vaultClient *vault.Client, vaultCo
 	return strings.HasPrefix(value, "vault:") || strings.HasPrefix(value, ">>vault:")
 }
+
+func hasInlineVaultDelimiters(value string) bool {
```
Sorry, I just checked that we use `${` `}` instead of the double `{{`.

https://github.com/banzaicloud/bank-vaults/blob/9c8215ecde27930d10e20605ea817cf2756cbf30/internal/configuration/template.go#L46

bonifaido commented on cmd/vault-secrets-webhook/configmap.go:
```diff
@@ -88,6 +101,28 @@ func (mw *mutatingWebhook) mutateConfigMapData(configMap *corev1.ConfigMap, data
 }
 
+func (mw *mutatingWebhook) mutateInlineConfigMapData(configMap *corev1.ConfigMap, data map[string]string, vaultClient *vault.Client, vaultConfig VaultConfig) error {
+	r, err := regexp.Compile("\\${{(.*?)}}")
```
We should compile the regex beforehand into "constant" variable I think.

bonifaido commented on cmd/vault-secrets-webhook/object.go:
```diff
@@ -95,6 +98,22 @@ func traverseObject(o interface{}, vaultClient *vault.Client, vaultConfig VaultC
 
 				e.Set(dataFromVault["data"])
+			} else if hasInlineVaultDelimiters(s) {
+				r, err := regexp.Compile("\\${{(.*?)}}")
```
And here as well, also this code part looks very much the same as for Secret.

kschu91 commented:
@bonifaido

I have extracted the matching logic into its own function to reduce duplicate code. However, I am not that familiar with go and have no idea how to do the same with the setting logic as there are different types of data coming in for configmaps, secrets and objects. But maybe this is good enough already?

Also, I globalized the compiliation of the regex, changed the behaviour to match only single brackets `${...}` and optimized the regex to match only single brackets containing `vault:` and `>>vault:`.

bonifaido reviewed (approved):
LGTM, the last thing it would be nice to have to add some kind of small test-case for this to make sure it doesn't break in the future: https://github.com/banzaicloud/bank-vaults/blob/master/hack/acceptance-test.sh#L137

bonifaido commented on hack/acceptance-test.sh:
```diff
@@ -140,11 +140,13 @@ helm upgrade --install vault-secrets-webhook ./charts/vault-secrets-webhook \
 kubectl apply -f deploy/test-secret.yaml
 test `kubectl get secrets sample-secret -o jsonpath='{.data.\.dockerconfigjson}' | base64 --decode | jq -r '.auths[].username'` = "dockerrepouser"
 test `kubectl get secrets sample-secret -o jsonpath='{.data.\.dockerconfigjson}' | base64 --decode | jq -r '.auths[].password'` = "dockerrepopassword"
+test `kubectl get secrets sample-secret -o jsonpath='{.data.inline}' | base64 --decode` = "Inline: secretId AWS_ACCESS_KEY_ID"
```
The first argument needs to be quoted I think otherwise thanks for the test!

```
++ kubectl get secrets sample-secret -o 'jsonpath={.data.inline}'
++ base64 --decode
+ test Inline: secretId AWS_ACCESS_KEY_ID = 'Inline: secretId AWS_ACCESS_KEY_ID'
hack/acceptance-test.sh: line 140: test: too many arguments
```

bonifaido commented on hack/acceptance-test.sh:
```diff
@@ -140,11 +140,13 @@ helm upgrade --install vault-secrets-webhook ./charts/vault-secrets-webhook \
 test `kubectl get cm sample-configmap -o jsonpath='{.data.aws-access-key-id}'` = "secretId"
 test "$(kubectl get cm sample-configmap -o jsonpath='{.data.aws-access-key-id-formatted}')" = "AWS key in base64: c2VjcmV0SWQ="
 test `kubectl get cm sample-configmap -o jsonpath='{.binaryData.aws-access-key-id-binary}'` = "secretId"
+test `kubectl get cm sample-configmap -o jsonpath='{.data.inline}'` = "Inline: secretId AWS_ACCESS_KEY_ID"
```
And here as well.

kschu91 commented:
> LGTM, the last thing it would be nice to have to add some kind of small test-case for this to make sure it doesn't break in the future: https://github.com/banzaicloud/bank-vaults/blob/master/hack/acceptance-test.sh#L137

@bonifaido We have updated the acceptance tests. Would be great if we can hava a merge along with a new release soon!

bonifaido commented:
No worries, a new release will performed tomorrow I think!

bonifaido commented:
https://github.com/banzaicloud/bank-vaults/pull/1157
